### PR TITLE
IndexWriterProjector support to convert to Iterator

### PR DESCRIPTION
**Pull request by andreidan**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (loading == null) {
+            return false;
+        }
+        raiseIfLoading();
+
+        if (currentRow.equals(OFF_ROW) && result == null) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: How can `result == null` ever be true here? There is a check couple of lines above which causes moveNext to return early

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    public CompletionStage<?> loadNextBatch() {
+        if (isLoading()) {
+            return CompletableFutures.failedFuture(new IllegalStateException("Iterator is already loading"));
+        }
+
+        if (allLoaded()) {
+            return CompletableFutures.failedFuture(new IllegalStateException("All batches already loaded"));
+        }
+        loading = new CompletableFuture();
+        bulkShardProcessor.result().whenComplete((r, t) -> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: Would it make sense to assign `bulkShardProcessor.result()` to `loading` - avoiding the intermediate future?

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        } catch (Exception e) {
+            loading.completeExceptionally(e);
+        }
+        return loading;
+    }
+
+    private void consumeSource() {
+        while (source.moveNext()) {
+            if (!consumeRow(source.currentRow())) {
+                return;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: Maybe add a comment clarifying how the future is triggered in this case

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java (hunk 2 above)
Comment (andreidan) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: I initially thought about that too, but I don't think it's a good idea to expose it to the client code. 

There's also a couple of shortcuts where we fail the loading future earlier, if the source yields an exception while consuming it for example.

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import io.crate.operation.collect.RowShardResolver;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.bulk.BulkShardProcessor;
+
+import java.util.BitSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+public class IndexWriterCountBatchIterator implements BatchIterator {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: Please add a short description explaining what this batchIterator does

Comment (dobe) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: pls document

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+    @Override
+    public boolean moveNext() {
+        if (loading == null) {
+            return false;
+        }
+        raiseIfLoading();
+
+        if (currentRow.equals(OFF_ROW) && result == null) {
+            result = loading.join();
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: can we prevent the pattern of using the future as result so we can get rid of this join.

**Pull request merged**